File: app/src/main/java/com/github/dedis/student20_pop/model/network/method/Subscribe.java
package com.github.dedis.student20_pop.model.network.method;

import com.github.dedis.student20_pop.utility.protocol.MessageHandler;

/**
 * Subscribe to a channel
 */
public final class Subscribe extends Query {

    public Subscribe(String channel, int id) {
        super(channel, id);
    }

    @Override
    public void accept(MessageHandler handler) {
        handler.handle(this);
    }

    @Override
    public String getMethod() {
        return Method.SUBSCRIBE.getMethod();
    }
}

File: be1-go/internal/handler/method/subscribe/subscribe.go
package subscribe

import (
	"encoding/json"
	"popstellar/internal/handler/messagedata/root"
	"popstellar/internal/repository"

	"popstellar/internal/errors"
	"popstellar/internal/message/query/method"
	"popstellar/internal/network/socket"
)

type Handler struct {
	subs repository.SubscriptionManager
}

func New(subs repository.SubscriptionManager) *Handler {
	return &Handler{subs: subs}
}

func (h *Handler) Handle(socket socket.Socket, msg []byte) (*int, error) {
	var subscribe method.Subscribe
	err := json.Unmarshal(msg, &subscribe)
	if err != nil {
		return nil, errors.NewJsonUnmarshalError(err.Error())
	}

	if root.Root == subscribe.Params.Channel {
		return &subscribe.ID, errors.NewAccessDeniedError("cannot Subscribe to root channel")
	}

	err = h.subs.Subscribe(subscribe.Params.Channel, socket)
	if err != nil {
		return &subscribe.ID, err
	}

	socket.SendResult(subscribe.ID, nil, nil)

	return &subscribe.ID, nil
}
